File: src/main/java/io/tripled/social/client/presentation/request/RequestDispatcher.java
package io.tripled.social.client.presentation.request;

import io.tripled.social.client.presentation.Output;
import io.tripled.social.client.presentation.controller.FollowController;
import io.tripled.social.client.presentation.controller.PostController;
import io.tripled.social.client.presentation.controller.ReadController;
import io.tripled.social.client.presentation.controller.WallController;

public class RequestDispatcher implements Visitor {
  private final PostController postController;
  private final ReadController readController;
  private final WallController wallController;
  private final FollowController followController;


  public RequestDispatcher(PostController postController,
                           ReadController readController,
                           WallController wallController,
                           FollowController followController) {
    this.postController = postController;
    this.readController = readController;
    this.wallController = wallController;
    this.followController = followController;
  }

  @Override
  public void visit(FollowCliRequest request, Output output) {
    followController.execute(request, output);
  }

  @Override
  public void visit(PostCliRequest request, Output output) {
    postController.execute(request, output);
  }

  @Override
  public void visit(ReadCliRequest request, Output output) {
    readController.execute(request, output);
  }

  @Override
  public void visit(WallCliRequest request, Output output) {
    wallController.execute(request, output);
  }
}

File: src/main/kotlin/io/tripled/social/client/presentation/request/RequestDispatcher.kt
package io.tripled.social.client.presentation.request

import io.tripled.social.client.presentation.Output
import io.tripled.social.client.presentation.controller.FollowController
import io.tripled.social.client.presentation.controller.PostController
import io.tripled.social.client.presentation.controller.ReadController
import io.tripled.social.client.presentation.controller.WallController

class RequestDispatcher(
    private val postController: PostController,
    private val readController: ReadController,
    private val wallController: WallController,
    private val followController: FollowController
) : Visitor {
    override fun visit(request: FollowCliRequest, output: Output) {
        followController.execute(request, output)
    }

    override fun visit(request: PostCliRequest, output: Output) {
        postController.execute(request, output)
    }

    override fun visit(request: ReadCliRequest, output: Output) {
        readController.execute(request, output)
    }

    override fun visit(request: WallCliRequest, output: Output) {
        wallController.execute(request, output)
    }
}